### DISCUSSION TITLE
Tests for MVIKotlin components

### DISCUSSION THREAD

**Discussion by DunzoLandroidlit** [arkivanov/MVIKotlin#86] [Q&A]
I am having a hard time writing tests for *Executor in my application which is extending CoroutineExecutor.
I want the ability to assert that certain messages were dispatched and certain labels were published based on intentions it receives. Please point me to any code snippet/ideas which will help me write these tests.

**Comment by arkivanov** [comment]
Thanks for the question, converted to a discussion.

Did you check samples? A `Store` is supposed to be test as a whole, and the `Executor` is designed to be implementation details. Here is an example of how you can verify Labels of a Store - [DetailsStoreTest.kt](https://github.com/arkivanov/MVIKotlin/blob/1c1326bc8ce0d503a7475d82c13a6c1ca6e8d373/sample/coroutines/shared/src/commonTest/kotlin/com/arkivanov/mvikotlin/sample/coroutines/shared/details/store/DetailsStoreTest.kt#L65).

But if you really want to test just `Executor`, the you will need to call its `init` method and supply an implementation of `Callbacks` interface. Override `Callbacks#onMessage` and `Callbacks#onLabel` and collect to a list. Then verify the connect of the list.

**Reply by arkivanov** [reply]
You can also create a test implementation of `Callbacks` for your tests. Something as follows should work.

```kotlin
class TestExecutorCallbacks<State, Message, Label>(
    override var state: State,
) : Executor.Callbacks<State, Message, Label> {

    val messages: MutableList<Message> = ArrayList()
    val labels: MutableList<Label> = ArrayList()


    override fun onMessage(message: Message) {
        messages += message
    }

    override fun onLabel(label: Label) {
        labels += label
    }
}
```

**Reply by DunzoLandroidlit** [reply]
This is the exact approach i followed, but there is a small caveat

`open class CoroutineExecutor<in Intent : Any, in Action : Any, in State : Any, Message : Any, Label : Any>(
    mainContext: CoroutineContext = Dispatchers.Main
) : Executor<Intent, Action, State, Message, Label> {

    private val callbacks = atomic<Callbacks<State, Message, Label>>()
    private val getState: () -> State = { callbacks.requireValue().state }

    /**
     * A [CoroutineScope] that can be used by the [CoroutineExecutor] descendants to launch coroutines.
     * The [CoroutineScope] is automatically cancelled on dispose.
     */
    protected val scope: CoroutineScope = CoroutineScope(mainContext)
.....`


This is the definition of CoroutineExecutor from the library. There is no way to use a TestCoroutineScope(mainContext) for the scope val which is defined within the class. This causes problems and i have to introduce a delay in the tests so that they pass the assertContains(*) assertions.

While i do agree on the way suggested to test entire Store object and put assertions on state and labels and pass intentions manually, we as an organisation are trying to push coverage for all components. Would be really helpful if some MVITest sort of library makes it easier to test things out. Cheers.

I might share a code snippet in some sample code if time permits in future.
Would appreciate if you point out if  i am missing something here. Cheers and thanks for the great lib.

**Reply by arkivanov** [reply]
You should use [TestCoroutineDispatcher](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-test-coroutine-dispatcher/) to test delays, etc. Just pass as `mainContext` value. Would this work?

**Reply by arkivanov** [reply]
Sorry, it appears that `TestCoroutineDispatcher` is deprecated. The new way is to use either [StandardTestDispatcher](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-standard-test-dispatcher.html) or [UnconfinedTestDispatcher](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-unconfined-test-dispatcher.html).
